tests/test_decoder.py

```
@@ -0,0 +1,482 @@
+                    expected_output = codec
+                self.detect_encoding_test_helper(
+                    string,
+                    string.encode(codec),
+                    expected_output
+                )
+
+    def auto_decode_test_helper(
+        self,
+        input_bytes,
+        codec_for_input_bytes,
+        expected_output
+    ):
+        ERROR = "auto_decode({}) returned the wrong value."
+        does_auto_detect_encodings_return_value_matter = (
+            codec_for_input_bytes is not None and (
+                encoding_detectable(expected_output, codec_for_input_bytes)
+                or len(input_bytes) == 0
+            )
+        )
+        if does_auto_detect_encodings_return_value_matter:
+            actual_output = decoder.auto_decode(input_bytes)
+            self.assertEqual(
+                expected_output,
+                actual_output,
+                msg=ERROR.format(repr(input_bytes))
+            )
+            self.assertIsInstance(actual_output, str)
+        else:
+            try:
+                decoder.auto_decode(input_bytes)
+            except UnicodeDecodeError as exception:
+                return exception
+        return None
+
+    def test_auto_decode_with_pre_encoded_strings(self):
+        ERROR = "auto_decode({}) should not have raised an exception"
+        for pre_encoded_test_string_info in PRE_ENCODED_TEST_STRING_INFOS:
+            exception = self.auto_decode_test_helper(
+                pre_encoded_test_string_info.input_bytes,
+                pre_encoded_test_string_info.codec_for_input_bytes,
+                pre_encoded_test_string_info.expected_output_str
+            )
+            if exception is not None:
+                new_exception = self.failureException(
+                    msg=ERROR.format(
+                        repr(pre_encoded_test_string_info.input_bytes)
+                    )
+                )
+                raise new_exception from exception
```

I understand the logic and I'm OK to keep this code, but if it was me I would simplify all this and just let the exception pop up from `auto_decode_test_helper()`. It's a less clean, but shorter :wink:

(same later)

---

Now that I look at it again, I think that it makes more sense for the `UnicodeDecodeError` to be raised instead of returned. I just pushed a new version that lets it get raised instead of returned.